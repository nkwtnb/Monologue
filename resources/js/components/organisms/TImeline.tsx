import React, { memo, useEffect, useLayoutEffect } from 'react';
import { useState } from 'react';
import Post from './Post';
import axios from 'axios';

interface Props {
  name: string | undefined;
  filter: "post" | "like";
}

interface Entry {
  name: string,
  id: number,
  avatar: string,
  words: string;
  created_at: string;
  likes: number;
  like: boolean;
}

export default (props: Props): JSX.Element => {

  const [entries, setEntries] = useState<Entry[]>([]);

  const handleLike = (e: any, id: number) => {
    let clickedEntry: Entry | undefined;
    for (let i=0; i<entries.length; i++) {
      const entry: any = entries[i];
      if (entry.id === id) {
        clickedEntry = entry;
        break;
      }
    }
    if (!clickedEntry) {
      return;
    }
    (async () => {
      if (clickedEntry.like) {
        const resp = await axios.delete("/api/likes", { data: { entryId: id } });
      } else {
        const resp = await axios.post("/api/likes", { entryId: id });
      }
      setEntries(() => {
        return entries.map(entry => {
          if (entry.id === id) {
            entry.like = !entry.like;
            entry.like ? (entry.likes++) : (entry.likes--);
          }
          return entry;
        })
      })
    })();
  }
  
  useEffect(() => {
    (async () => {
      //エントリー取得
      let resp;
      if (props.name === "") {
        resp = await axios.get(`/words`);
      } else {
        let filterString = "";
        if (props.filter === "like") {
          filterString = "/?filter=" + props.filter;
        }
        resp = await axios.get(`/words/${props.name}${filterString}`);
      }
      const respLikes = await axios.get("/likes");
      const _likes = respLikes.data;
      const _entries = resp.data;
      const mapped = _entries.map((entry: any) => {
        entry.like = _likes.indexOf(entry.id) > -1 ? true : false;
        return entry;
      });
      setEntries(mapped);
    })();
  }, [props.name, props.filter]);

  return (
    <>
      {
        entries.map((entry: Entry, index) => {
          return (
            <div key={index}>
              <Post
                id={entry.id}
                name={entry.name}
                created_at={entry.created_at}
                avatar={entry.avatar}
                words={entry.words}
                like={entry.like}
                likes={entry.likes}
                handleLike={handleLike}
              />
            </div>
          )
        })
      }
    </>
  );
}