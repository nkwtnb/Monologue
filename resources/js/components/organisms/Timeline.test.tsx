import React from "react";
import TImeline from "./TImeline";
import { cleanup, render, screen } from "@testing-library/react";

afterEach(() => cleanup());
describe("Timeline unit test", () => {
  it("配列で受け取ったメッセージが表示される", () => {
    const result = render(
      <TImeline filter="post" name="test"></TImeline>
    );
    console.log(screen.debug());
    expect(result.asFragment()).toMatchSnapshot();
  });
})