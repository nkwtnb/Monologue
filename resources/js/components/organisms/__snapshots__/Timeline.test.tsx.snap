// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline unit test 配列で受け取ったメッセージが表示される 1`] = `[Function]`;
